pootle/apps/virtualfolder/utils.py

```diff
@@ -0,0 +1,124 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) Pootle contributors.
+#
+# This file is a part of the Pootle project. It is distributed under the GPL3
+# or later license. See the LICENSE file for a copy of the license and the
+# AUTHORS file for copyright and authorship information.
+
+import os
+
+from pootle.core.url_helpers import split_pootle_path
+from pootle_store.models import Store
+
+
+def split_vfolder_location(path):
+    (lang_code, proj_code,
+     dir_path, filename) = split_pootle_path(path)
+    if lang_code == "{LANG}":
+        lang_code = None
+    if proj_code == "{PROJ}":
+        proj_code = None
+    return lang_code, proj_code, dir_path, filename
+
+
+class VirtualFolderStoreMatcher(object):
+
+    def __init__(self, vf):
+        self.vf = vf
+
+    def add_related_stores(self):
+        existing_stores = self.vf.stores.all()
+        related_stores = self.find_related_stores()
+        for store in related_stores:
+            if store not in existing_stores:
+                self.vf.stores.add(store)
+
+    def add_store_if_related(self, store):
+        if store in self.find_related_stores():
+            self.vf.stores.add(store)
+
+    @property
+    def filter_rules(self):
+        return (
+            r.strip()
+            for r
+            in self.vf.filter_rules.split(","))
+
+    @property
+    def path_parts(self):
+        return split_vfolder_location(self.vf.location)
+
+    @property
+    def lang_code(self):
+        return self.path_parts[0]
+
+    @property
+    def proj_code(self):
+        return self.path_parts[1]
+
+    @property
+    def store_qs(self):
+        qs = Store.objects
+        if self.lang_code:
+            qs = qs.filter(translation_project__language__code=self.lang_code)
+        if self.proj_code:
+            qs = qs.filter(translation_project__project__code=self.proj_code)
+        if self.use_regex:
+            if self.lang_code or self.proj_code:
+                qs = qs.filter(pootle_path__regex="^%s" % self.path)
+        else:
+            qs = qs.filter(pootle_path__startswith=self.path)
+        return qs
+
+    @property
+    def use_regex(self):
+        return (
+            not self.lang_code
+            or not self.proj_code)
+
+    @property
+    def path_regex(self):
+        regex = "/[^/]*"
+        return self.vf.location.replace(
+            "/{LANG}", regex).replace("/{PROJ}", regex)
+
+    @property
+    def path(self):
+        return (
+            self.path_regex
+            if self.use_regex
+            else self.vf.location)
+
+    def get_rule_path(self, rule):
+        return "%s%s" % (self.path, rule)
+
+    def filter_by_rule(self, qs, rule):
+        rule_path = self.get_rule_path(rule)
+        is_file = "." in os.path.basename(rule)
+        if self.use_regex:
+            return (
+                qs.filter(pootle_path__regex="^%s$" % rule_path)
+                if is_file
+                else qs.filter(pootle_path__regex="^%s" % rule_path))
```

Comment:
i would imagine so - its the equiv of difference between endswith and matches - but on a more complex query